File: python/numprocs.py
#!/usr/bin/env python
# Check that it is not possible to allocate more resources than requested at allocation time
import subprocess as sp
# just using some random numbers with tasks requested < tasks to execute
p = sp.Popen(['salloc', '--ntasks=2'], stdin=sp.PIPE, stderr=sp.PIPE, stdout=sp.PIPE)
# out = (_,_), because the request is supposed to return an error tuple[0] will be empty
out = p.communicate(b'srun --ntasks=3 hostname')
if len(out[0]) == 0:
    print('PASS')
else:
    print('FAIL')

File: rust/slurm-test/rust/slurm-test/src/bin/numprocs.rs
use subprocess::{Popen, PopenConfig, Redirection, Result};
/// Verify that it is not possible to use more resources than originally
/// requested
fn main() -> Result<()> {
    let mut p = Popen::create(
        &["salloc", "--mem=1GB", "--ntasks=2"],
        PopenConfig {
            stdout: Redirection::Pipe,
            stdin: Redirection::Pipe,
            stderr: Redirection::Pipe,
            ..Default::default()
        },
    )?;
    let (_out, err) = p.communicate(Some("srun --ntasks=3 hostname"))?;
    if let Some(e) = err {
        if e.to_lowercase()
            .contains(&"more processors requested than permitted")
        {
            println!("PASS");
        } else {
            println!("FAIL");
        }
    } else {
        println!("FAIL");
    }
    Ok(())
}
